src/lib.rs

```diff
@@ -0,0 +1,284 @@
+// Copyright 2019 King's College London.
+// Created by the Software Development Team <http://soft-dev.org/>.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+/// YkMir -- Metadata serialiser and deserialiser for Yorick.
+///
+/// This crate allows ykrustc to serialise the compiler's MIR structures for later deserialisation
+/// by the Yorick runtime. The encoded data is specialised to a particular Rust input program (this
+/// is in contrast to rustc's notion of crate metadata which is stored in a generic form, so that
+/// later compilation sessions can perform their own specialisations). By storing specialised MIR,
+/// we avoid the need to carry a whole instance of the compiler with us at runtime to resolve
+/// generic metadata into specialised metadata (Rust's metadata loader API would require us to pass
+/// a `TyCtxt`).
+///
+/// The encoder and decoder API is structured in such a way that MIRs can be streamed in and out one
+/// at a time. This helps to reduce memory consumption in the compiler, avoiding the need to have
+/// an entire second version of the compiler's metadata in memory at compile-time.
+///
+/// The MIR data is serialised to msgpack data in the following form:
+///
+///  version            -- The ABI version number of the serialised data.
```

Comment:
Is `ABI` the right term here?

Reply:
Ah hah! I wasn't sure myself.

The version is just a number we bump when shape of the data structures serialised within. For example, when I add statement support I would bump the number, as `struct Mir` will get bigger and we don't want an earlier deserialiser to try and decode it.

What would be the correct term? "serialisation version"?

Reply:
"Serialisation version" works for me.